src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionSource_ISuggestedActionsSourceExperimental.cs

```diff
@@ -0,0 +1,123 @@
+                ReferenceCountedDisposable<State> state,
+                ISuggestedActionCategorySet requestedActionCategories,
+                Document document,
+                SnapshotSpan range,
+                Func<string, IDisposable?> addOperationScope,
+                CodeActionProviderPriority priority,
+                [EnumeratorCancellation] CancellationToken cancellationToken)
+            {
+                var workspace = document.Project.Solution.Workspace;
+                var supportsFeatureService = workspace.Services.GetRequiredService<ITextBufferSupportsFeatureService>();
+
+                var selection = TryGetCodeRefactoringSelection(state, range);
+
+                var fixesTask = GetCodeFixesAsync(
+                    state, supportsFeatureService, requestedActionCategories, workspace, document, range,
+                    addOperationScope, priority, isBlocking: false, cancellationToken);
+                var refactoringsTask = GetRefactoringsAsync(
+                    state, supportsFeatureService, requestedActionCategories, workspace, document, selection,
+                    addOperationScope, priority, isBlocking: false, cancellationToken);
+
+                if (priority == CodeActionProviderPriority.High)
+                {
+                    // in a high pri scenario, return data as soon as possible so that the user can interact with them.
+                    // this is especially important for state-machine oriented refactorings (like rename) where the user
+                    // should always have access to them effectively synchronously.
+                    var firstTask = await Task.WhenAny(fixesTask, refactoringsTask).ConfigureAwait(false);
+                    var secondTask = firstTask == fixesTask ? refactoringsTask : fixesTask;
+
+                    var orderedTasks = new[] { firstTask, secondTask };
+                    foreach (var task in orderedTasks)
+                    {
+                        if (task == fixesTask)
+                        {
+                            var fixes = await fixesTask.ConfigureAwait(false);
+                            foreach (var set in ConvertToSuggestedActionSets(state, selection, fixes, ImmutableArray<UnifiedSuggestedActionSet>.Empty))
+                                yield return set;
+                        }
+                        else
+                        {
+                            Contract.ThrowIfFalse(task == refactoringsTask);
+
+                            var refactorings = await refactoringsTask.ConfigureAwait(false);
+                            foreach (var set in ConvertToSuggestedActionSets(state, selection, ImmutableArray<UnifiedSuggestedActionSet>.Empty, refactorings))
+                                yield return set;
+                        }
+                    }
+                }
+                else
+                {
+                    await Task.WhenAll(fixesTask, refactoringsTask).ConfigureAwait(false);
```

Comment:
What is the WhenAll really doing here? You could delete this and it'd all work the same, right?

Reply:
(my assumption was either we'd use the result from WhenAll, or we'd do .Result on the tasks since we know they're then complete or something. It made me wonder if I had mis-read the code and the tasks on line 114 and on 115-116 were actually different.)

Reply:
agreed.  fixed.